ImputationPipeline/ScoreBGE/ScoreBGE.py

```diff
@@ -0,0 +1,266 @@
+
+        self.vcf_sample_score = {sample_name: 0 for sample_name in self.sample_names}
+        self.vcf_sites_scored = {sample_name: [] for sample_name in self.sample_names}
+        self.vcf_num_sites_not_found = 0
+
+        for _, weight in self.prs_weights.iterrows():
+            site_records = vcf.fetch(weight.contig, weight.position - 1, weight.position)
+
+            # Get first record
+            record = next(site_records, None)
+            self._check_biallelic(record)
+
+            # Find the record that matches the weight
+            while record is None or record.pos != weight.position or record.ref != weight.ref or weight.alt != record.alts[0]:
+                # If fetch yields no more records at this position, break here
+                if record is None:
+                    break
+                record = next(site_records, None)
+                self._check_biallelic(record)
+
+            # Skip the weight if the record was not found
+            if record is None:
+                self.vcf_num_sites_not_found += 1
+                print('  Site not found in VCF:', weight.locus, weight.ref, weight.alt)
+                continue
+            
+            self._vcf_score_dosage(record, weight)
+        self._print_wgs_vcf_metrics()
+        if self.gvcf_sites_scored is not None:
+            self._print_wes_and_wgs_metrics()
+
+    def write_output(self, basename, allow_single_source_scoring=False):
+        """
+        Write the scores to plink-like output files.
+        """
+
+        # GVCF
+        if self.gvcf_sites_scored_set is None:
+            if not allow_single_source_scoring:
+                raise RuntimeError('No GVCF scoring performed. If you want to output scores from a single source, set allow_single_source_scoring to True.')
+            else:
+                print('No GVCF scoring performed. Skipping GVCF output.')
+        else:
+            with open(f'{basename}.exome_gvcf.score', 'w') as out_exome_gvcf_score:
+                out_exome_gvcf_score.write('#IID\tSCORE1_SUM\n')
+                for sample_name in self.sample_names:
+                    out_exome_gvcf_score.write(f'{sample_name}\t{self.gvcf_sample_score[sample_name]}\n')
+
+            with open(f'{basename}.exome_gvcf.sites_scored', 'w') as out_exome_gvcf_sites_scored:
+                out_exome_gvcf_sites_scored.write('site	samples_scored\n')
+                for weight in self.prs_weights.iterrows():
+                    locus = weight[1]['locus']
+                    ref = weight[1]['ref']
+                    alt = weight[1]['alt']
+                    weight_samples = [sample_name for sample_name in self.sample_names if (locus, ref, alt) in self.gvcf_sites_scored[sample_name]]
+                    out_exome_gvcf_sites_scored.write(f'{locus}:{ref}:{alt}\t{",".join(weight_samples)}\n')
+
+        # VCF
+        if self.vcf_sites_scored is None:
+            if not allow_single_source_scoring:
```

Comment:
could refactor this a little, since it's pretty identical to the gvcf code, but not a big deal either way.